### PR TITLE
Decrypt only once and hold calendar auth data in memory

### LINKED ISSUE #88: Runtime is too long
CalendarSync is kinda slow until the first browser window opens. Maybe there's some room for improvement.

### PR DESCRIPTION
**Pull request by alxndr13**
fixes #88 and #29

### CODE REVIEW

**Comment by alxndr13** [comment]
Probably a bit rushed, but works :)

File: internal/auth/yaml_storage.go [hunk 1]
```diff
@@ -47,10 +59,25 @@ func (y *YamlStorage) WriteCalendarAuth(newCal CalendarAuth) (bool, error) {
 		return false, err
 	}
 
+	// Adding freshly written CalendarAuth to memory
+	// Probably unneeded, the next time this data will be retrieved is on the next calendarsync run
+	log.Debugf("Adding calendar auth for cal %s to memory", newCal.CalendarID)
+	y.DecryptedAuth = append(y.DecryptedAuth, newCal)
```
Comment (MichaelEischer) [inline internal/auth/yaml_storage.go:65]: I'd simplify this to `y.DecryptedAuth = cals`. That trivially ensures that the cached data and the on-disk state are equivalent.

File: internal/auth/yaml_storage.go [hunk 2]
```diff
@@ -8,18 +8,30 @@ import (
...
+	// Holds the decrypted CalendarAuth Config in memory, so the file does not have to be read multiple times
+	DecryptedAuth []CalendarAuth
 }
 
 func (y *YamlStorage) Setup(config config.AuthStorage, encryptionPassphrase string) error {
 	y.StorageEncryptionKey = encryptionPassphrase
 	y.StoragePath = config.Config["path"].(string)
+
+	log.Debug("Loading saved auth data into memory")
+	stor, err := y.readAndParseFile()
```
Comment (MichaelEischer) [inline internal/auth/yaml_storage.go:28]: What about changing this to lazy loading? Then `ReadCalendarAuth()` would store the decrypted data after loading it. That would also fix the error handling if err is not an `os.ErrNotExist`

File: internal/auth/yaml_storage.go [hunk 3]
```diff
@@ -8,18 +8,30 @@ import (
...
+	"github.com/charmbracelet/log"
 	"github.com/inovex/CalendarSync/internal/config"
 	"gopkg.in/yaml.v3"
 )
 
 type YamlStorage struct {
 	StoragePath          string
 	StorageEncryptionKey string
+	// Holds the decrypted CalendarAuth Config in memory, so the file does not have to be read multiple times
+	DecryptedAuth []CalendarAuth
```
Comment (MichaelEischer) [inline internal/auth/yaml_storage.go:20]: (Nitpick) I'd slightly prefer `CachedAuth`, although I'm also ok with keeping the current name.

File: internal/auth/yaml_storage.go [hunk 4]
```diff
@@ -47,10 +59,25 @@ func (y *YamlStorage) WriteCalendarAuth(newCal CalendarAuth) (bool, error) {
...
+	// Probably unneeded, the next time this data will be retrieved is on the next calendarsync run
+	log.Debugf("Adding calendar auth for cal %s to memory", newCal.CalendarID)
+	y.DecryptedAuth = append(y.DecryptedAuth, newCal)
+
 	return true, nil
 }
 
 func (y *YamlStorage) ReadCalendarAuth(calendarID string) (*CalendarAuth, error) {
+	// if we already decrypted the file, read from memory
+	if len(y.DecryptedAuth) > 0 {
```
Comment (MichaelEischer) [inline internal/auth/yaml_storage.go:72]: There's no need to duplicate the calendarID filtering:
```
var calendars []CalendarAuth
if len(y.DecryptedAuth) > 0 {
    calendars = y.DecryptedAuth
} else {
    // readAndParseFile
}
for _, cal := range calendars {
```

**Review by MichaelEischer — COMMENTED** [review]
The code looks like it works, although I'd prefer to clean it up a bit.

File: internal/auth/yaml_storage.go (hunk 1 above)
Comment (alxndr13) [reply internal/auth/yaml_storage.go:65]: fixed, good catch!

File: internal/auth/yaml_storage.go (hunk 4 above)
Comment (alxndr13) [reply internal/auth/yaml_storage.go]: fixed :)

File: internal/auth/yaml_storage.go (hunk 3 above)
Comment (alxndr13) [reply internal/auth/yaml_storage.go]: CachedAuth is a way better name imo. Renamed

File: internal/auth/yaml_storage.go (hunk 2 above)
Comment (alxndr13) [reply internal/auth/yaml_storage.go]: i had it like this, but due to my mechanism of filtering in the "ReadCalendarAuth" func the file was loaded twice. (as i returned just the auth data for the specific calendar id)

fixed.

**Review by MichaelEischer — APPROVED** [review]

**Pull request merged**